luks2/container.go

```diff
@@ -0,0 +1,233 @@
+ *
+ */
+
+package luks2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/secboot"
+	"github.com/snapcore/secboot/internal/luks2"
+	"golang.org/x/sys/unix"
+)
+
+// StorageContainer represents a LUKS2 storage container.
+type StorageContainer interface {
+	secboot.StorageContainer
+
+	// Dev returns the device number for the LUKS2 storage container.
+	// The major and minor components of this can be accessed with
+	// unix.Major and unix.Minor.
+	Dev() uint64
+
+	// ActiveVolumeName returns the associated volume name if this
+	// StorageContainer is active. If it isn't active, it is
+	// expected to return a secboot.ErrStorageContainerNotActive
+	// error.
+	ActiveVolumeName(ctx context.Context) (string, error)
+
+	// TODO: Add LUKS2 UUID / label accessors here as well.
+}
+
+type activateOptionKey string
+
+const (
+	volumeNameKey activateOptionKey = "volume-name"
+)
+
+type activateOptions map[any]any
+
+func (o activateOptions) Add(key, value any) {
+	o[key] = value
+}
+
+type volumeNameOption string
+
+func (o volumeNameOption) ApplyTo(visitor secboot.ActivateOptionVisitor) {
+	if _, ok := visitor.(activateOptions); !ok {
+		panic("WithVolumeName can only be used for the luks2 backend")
+	}
+	visitor.Add(volumeNameKey, string(o))
+}
+
+// WithVolumeName is used to specify the volume name used to create the
+// DM device when unlocking a LUKS2 container.
+//
+// XXX: It is mandatory for now for any [StorageContainer] managed by this
+// package. A future revision will relax this requirement and support a
+// mode where the volume name is created automatically, perhaps based on
+// the UUID of the LUKS2 container. An automatically generated volume name
+// can be retrieved afterwards using [StorageContainer.ActiveVolumeName].
+func WithVolumeName(name string) secboot.ActivateOption {
+	return volumeNameOption(name)
+}
+
+// storageContainerImpl is an implementation of [secboot.StorageContainer].
+type storageContainerImpl struct {
+	// path is the canonical path of the device node for this
+	// storage container.
+	path string
+
+	// dev is the unix device number for this storage container.
+	dev uint64
+}
+
+// Dev implements [StorageContainer.Dev].
+func (c *storageContainerImpl) Dev() uint64 {
+	return c.dev
+}
+
+// ActiveVolumeName implements [StorageContainer.ActiveVolumeName].
+func (c *storageContainerImpl) ActiveVolumeName(ctx context.Context) (string, error) {
+	// To determine if this container is associated with an active volume,
+	// and in order to obtain the volume name, we need to iterate over
+	// the existing device mapper volumes to find any that are backed by
+	// this storage container.
+	dmDevices, err := filepath.Glob(filepath.Join(devRoot, "dm-*"))
+	if err != nil {
+		return "", fmt.Errorf("cannot build list of active dm devices: %w", err)
+	}
+
+	// Iterate over the list of active DM volumes.
+	for _, path := range dmDevices {
```

Comment:
Maybe we could make it look at only the holders of the device?

Reply:
I wonder if we could do that and skip the call to `sourceDeviceFromDMDevice`. For that to work, we would need to ensure that there is only ever a single DM device holding the top-level block device. I checked on my classic install which uses the stacked LVM on LUKS2 setup and the nested DM device for the LVM volume isn't a holder of the top-level block device, eg:
```
$ ls -l /sys/dev/block/259\:3/holders/
total 0
lrwxrwxrwx 1 root root 0 Sep  4 16:33 dm-0 -> ../../../../../../../../../../virtual/block/dm-0
$ ls -l /sys/dev/block/259\:3/holders/dm-0/holders/
total 0
lrwxrwxrwx 1 root root 0 Sep  4 16:34 dm-1 -> ../../dm-1
$ ls -l /sys/dev/block/259\:3/holders/dm-0/holders/dm-1/holders/
total 0
```

I wonder if we can always rely on this to be true?

Reply:
We cannot totally rely on holders. Some device mapper types have multiple devices (verity, integrity, etc.), so we need to figure out which is which. But I think we can rely on holders to be there to trim down what we look at, instead of scanning for all devices.
Could you add a TODO or something here to mention that there might be better ways? And we can maybe optimize things in a later PR.

Reply:
Sure, that sounds good to me.